pyop2/fusion/filters.py

```diff
@@ -0,0 +1,133 @@
+#       derived from this software without specific prior written
+#       permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTERS
+# ''AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""Classes for handling duplicate arguments in parallel loops and kernels."""
+
+from collections import OrderedDict
+
+from pyop2.base import READ, RW, WRITE
+from pyop2.utils import flatten
+
+from coffee.utils import ast_make_alias
+from coffee import base as ast
+
+
+class Filter(object):
+
+    def _key(self, arg):
+        """Arguments accessing the same :class:`base.Dat` with the same
+        :class:`base.Map` are considered identical."""
+        if arg._is_dat:
+            return (arg.data, arg.map)
+        elif arg._is_mat:
+            return (arg.data,) + tuple(arg.map)
+        else:
+            return (arg.data,)
+
+    def loop_args(self, loops):
+        """Merge and return identical :class:`base.Arg`s appearing in ``loops``.
+        Merging two :class:`base.Arg`s means discarding duplicates and taking the
+        set union of the access modes (if Arg1 accesses Dat1 in READ mode and Arg2
+        accesses Dat1 in WRITE mode, then a single argument is returned with
+        access mode RW). Uniqueness is determined by ``self._key``."""
+
+        loop_args = [loop.args for loop in loops]
+        filtered_args = OrderedDict()
+        for args in loop_args:
+            for a in args:
+                fa = filtered_args.setdefault(self._key(a), a)
+                if a.access != fa.access:
+                    if READ in [a.access, fa.access]:
+                        # If a READ and some sort of write (MIN, MAX, RW, WRITE,
+                        # INC), then the access mode becomes RW
+                        fa.access = RW
+                    elif WRITE in [a.access, fa.access]:
+                        # Can't be a READ, so just stick to WRITE regardless of what
+                        # the other access mode is
+                        fa.access = WRITE
+                    else:
+                        # Neither READ nor WRITE, so access modes are some
+                        # combinations of RW, INC, MIN, MAX. For simplicity,
+                        # just make it RW.
+                        fa.access = RW
```

Comment:
This I did in a previous PR. It should be safe, as ``RW`` is more restrictive than ``INC`` (you're simply losing associativity and commutativity). Here you've the same ``Dat`` accessed at least twice and only once with ``INC``: no matter what the other access does (surely not a ``READ``, surely not a ``WRITE``), but those two properties go lost, so the safest thing to do I can think of is to set it to ``RW``